src/read.rs

```diff
@@ -301,45 +299,45 @@ impl<R: Read+io::Seek> ZipArchive<R>
     }
 }
 
-fn central_header_to_zip_file<R: Read+io::Seek>(reader: &mut R, archive_offset: u64) -> ZipResult<ZipFileData>
+fn central_header_to_zip_file(bytes: &mut ByteBuf, archive_offset: u64) -> ZipResult<ZipFileData>
 {
     // Parse central header
-    let signature = reader.read_u32::<LittleEndian>()?;
+    let signature = bytes.read_u32::<LittleEndian>()?;
     if signature != spec::CENTRAL_DIRECTORY_HEADER_SIGNATURE
     {
         return Err(ZipError::InvalidArchive("Invalid Central Directory header"))
     }
 
-    let version_made_by = reader.read_u16::<LittleEndian>()?;
-    let _version_to_extract = reader.read_u16::<LittleEndian>()?;
-    let flags = reader.read_u16::<LittleEndian>()?;
+    let version_made_by = bytes.get_u16_le();
```

Comment:
What if the central directory header lied and the size it reports is too small? Then the central directory buffer might end in the middle of a record, and one of these `get_*` will panic. The original `read_*` would have read garbage (bits of the end-of-file record) or return `Err(UnexpectedEof)`

Since you already gave it a `Read` impl and still use the original `read_u32` on line 305, maybe keep the `read_*` here as well?

Reply:
That would be possible, but then we're still left with the calls to `split_to` for which `Read` doesn't have an  equivalent. Maybe the way to do it would be to check the length at the start of the method, and return a `Err(UnexpectedEof)` if it is too short.

Reply:
>That would be possible, but then we're still left with the calls to `split_to` for which `Read` doesn't have an equivalent.

Yes, you'd still have to check `.remaining()` manually for the three variable-length fields before `split_to()`ing them.

>Maybe the way to do it would be to check the length at the start of the method

Yes, this would be the most efficient (avoids the bound checks), though you'd still need to check the three variable-length fields separately from the constant-sized header.

Reply:
Yep, cool, that's what I meant. I'll go and make these changes!